decidim/decidim
# Replace the WYSIWYG editor with TipTap

#### :tophat: What? Why?
This PR implements the TipTap editor for Decidim.

This is feature complete and works both in the legacy design (admin panel) as well as in the redesign (you need to test it locally not to break staging, instructions below).

Can be tested at (use default seed credentials):
https://tiptap.decidimqa.mainiotech.fi/

#### :pushpin: Related Issues
- Related to
  * #7289
  * #7585
  * quilljs/quill#3359
- Fixes #10082
- Fixes #9398

Here are also some issues we have had in the past with Quill that should be confirmed to be fixed when testing this:

- [x] When saving and re-editing the content, sub-lists are loaded and displayed correctly
  #9866
- [x] Character counter does not count the HTML tags in the character count, see e.g. new proposal
  #9821
- [x] Copy-pasting content to the editor works
  #7180
  * [x] Works also for description field of assembly meetings, i.e. editors with hashtags
    #4873
- [x] Bold text is not lost when copy-pasting content from another editor
  #8982
- [x] Editor loads the embedded video content when that is the only content
  #8331
  #9629
- [x] Editor loads correctly when it contains paragraphs that only have a single line break inside them
  #7896
- [x] File upload works through the toolbar as well as copy-paste
  #7712
- [x] Does not create a trailing paragraph after a video element
  #3067
- [x] Videos display correctly when added to the pages
  #2130
  * [x] Note that you need to have accepted cookies for this to work
    #9170 (note: [not yet implemented under redesign](https://github.com/decidim/decidim/issues/9753#issuecomment-1369056254))
- [x] Pasting plain text content that has two line breaks between two text lines creates two separate paragraphs
  #2000
  * E.g. `First paragraph\n\nSecond paragraph`
- [x] Hashtags are recognized at the beginning of the content
  #9616
- [x] External link indicator is not added for content inside the editor
  #9095
- [x] Editor adds `nofollow`, `noopener` and `noreferrer` for participant entered content
  #9047
  #8779
  #5651
- [x] Character counter counts the emoji characters correctly, the same way as backend
  #8916
- [x] Shift+Enter makes a single line break instead of another paragraph
  #7877
- [x] External links are displayed as inline content when the WYSIWYG content is shown
  #7280
- [x] Alt tags are not removed for existing rich text content that has images with alt tags
  #6920

#### Testing
- Clone this branch, create the development app (you may need to regenerate it as `packages` have changed)
- In another window, open [try.decidim.org](https://try.decidim.org/)
- Go to admin panel in both versions
- Find a view with WYSIWYG editor
- Try it out
- Compare that we have all functionality available that we used to have with Quill

Currently there does not seem to be any redesigned view yet with the editor (e.g. new proposal / new meeting / new initiative / etc.) but you can test it locally under the redesign by adding it to some page, e.g. at [the static page view](https://github.com/decidim/decidim/blob/develop/decidim-core/app/views/decidim/pages/_tabbed.html.erb) as follows:

```erb
          
          <div class="mt-6">
            <%= decidim_form_for(Decidim::StaticPage.new, url: "") do |form| %>
              <%= form.editor(:content) %>
            <% end %>
          </div>
```

And if you want to test the full toolbar with the video embed and image embed functionality, replace the line with `form.editor` with this (note that you need to be logged in for the image embedding to work because of the permissions set for the image upload endpoint):

```erb
<%= form.editor(:content, toolbar: "full") %>
```

Note that this PR already includes E2E tests (system tests) for testing the editor functionality under both, the legacy design as well as the redesign.

##### Extra features

The new editor also comes with some extra features that you can test:

- Automatically formatting pasted content
  * Try pasting a YouTube URL, e.g. `https://www.youtube.com/watch?v=f6JMgJAQ2tc`
  * Try pasting markdown content, e.g. `**this should appear as bold**`
- Markdown shortcuts
  * Try writing e.g. a heading to the editor, start the text with `##` + `SPACE` and continue writing your heading
  * Try writing e.g. a bold to the editor, write double stars `**` followed by some text and then again double stars `**`
- Character counter
  * You can try it at the participant side, e.g. when submitting a new proposal
- Hastags support, try e.g. with new proposal
- Mentions support (right now we do not have any WYSIWYG editor in Decidim that supports mentions but the support is built to the editor)
- Double clicking existing embedded content (images or videos) to edit them

### :camera: Screenshots

#### Participant editor with demonstration of the available controls

![Participant WYSIWYG](https://user-images.githubusercontent.com/864340/216444586-9b907fc6-2987-459b-860f-00b52ee774ae.png)

#### Admin editor with extra controls

![Admin editor view](https://user-images.githubusercontent.com/864340/216444750-6e87c3d6-3e90-453f-a53e-214fc784231a.png)

##### Link dialog

![WYSIWYG link dialog](https://user-images.githubusercontent.com/864340/216444829-385c8d89-8e09-4bf8-b384-2b96236ef7c9.png)

##### Image dialog

![WYSIWYG image dialog](https://user-images.githubusercontent.com/864340/216444942-743992b1-a985-42f9-8211-fb1e5988427b.png)

##### Video dialog

![WYSIWYG video dialog](https://user-images.githubusercontent.com/864340/216445216-3a6d27a4-9a87-49d1-955e-0097453555a7.png)

Supports YouTube and Vimeo by default or any arbitrary service as long as they provide embeddable video URLs.

##### Image resizing support

As in the old editor, we have image resizing support also in the new editor.

![WYSIWYG image resizing](https://user-images.githubusercontent.com/864340/216450341-3bc71106-1fd5-41ee-b67e-ae6a3a880618.png)

#### Redesigned editor

![Redesigned WYSIWYG](https://user-images.githubusercontent.com/864340/216447470-c99b51e3-e963-4cb6-976f-ab3bb1b63d5f.png)

##### Link dialog (redesigned)

![WYSIWYG link dialog (redesigned)](https://user-images.githubusercontent.com/864340/216447654-df0c3513-b723-42d1-9a3e-96e6f7bbe723.png)

##### Image dialog (redesigned)

![WYSIWYG image dialog (redesigned)](https://user-images.githubusercontent.com/864340/216448121-1af36afd-d38f-4e82-8fae-87e83ff73c44.png)

##### Video dialog (redesigned)

![WYSIWYG video dialog (redesigned)](https://user-images.githubusercontent.com/864340/216448011-79ee3305-6253-4f95-a08f-de70bb93075f.png)
